Implement ask_for_reset for HTMLSelectElement.

dagnir opened the pull request:
Fixes #7774

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/7963)

<!-- Reviewable:end -->

dagnir committed:
Implement ask_for_reset for HTMLSelectElement.

Fixes #7774

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @SimonSapin (or someone else) soon.

frewsxcv commented:
`ask_for_reset` (the function you just created) should get called [here](https://github.com/servo/servo/blob/babeed998d297d2d7badc06ba9c8f90e42239b4f/components/script/dom/htmloptionelement.rs#L147-L148)

dagnir commented:
Hmmm `ask_for_reset` calls `HTMLOptionElement::SetSeledcted` so that will make them mutually recursive.  Is this intentional?  Or should I be setting `HTMLOptionElement::selectedness` directly somehow?
<!-- Reviewable comment -K0IxHmM4AkNFkvT1W94:376709812 -->

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
> Or should I be setting HTMLOptionElement::selectedness directly somehow?

You could add a `set_selectedness` method on `HTMLOptionElement` so that you could set it directly
<!-- Reviewable comment -K0JHhZbBzCHmR-mHGFE:-1368268347 -->

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmloptionelement.rs, line 153 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JWjJEt3G-9f_4Jt7A)** ([raw file](https://github.com/servo/servo/blob/4599d424a6b4fc8665b83d635dc36dfee6a4fd2d/components/script/dom/htmloptionelement.rs#L153)):</sup>
Instead of doing `.find(|a| a.is_htmlselectelement)` and then `HTMLSelectElementCast::to_root(..).unwrap`, use `filter_map` with `HTMLSelectElementCast::to_root`

http://doc.servo.org/std/iter/trait.Iterator.html#method.filter_map

Let me know if you need help 
<!-- Reviewable comment -K0JWjJEt3G-9f_4Jt7B:1956109801 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

dagnir commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[components/script/dom/htmloptionelement.rs, line 153 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JWjJEt3G-9f_4Jt7A)** ([raw file](https://github.com/servo/servo/blob/4599d424a6b4fc8665b83d635dc36dfee6a4fd2d/components/script/dom/htmloptionelement.rs#L153)):</sup>
Thanks, changed to `filter_map`.
<!-- Reviewable comment -K0J_wfU1NESidUmBs52:838608428 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 78 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JoQIP2vSwpPy3Fr_I)** ([raw file](https://github.com/servo/servo/blob/3f529b0025bd01c4fa82defbd147dab8d9cb0a68/components/script/dom/htmlselectelement.rs#L78)):</sup>
This could be simplified to `is_none` and `get_enabled_state`
<!-- Reviewable comment -K0JoQIQYnnlmtBDkj4M:875183330 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 72 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JrEWd6-U2Hfn1thln)** ([raw file](https://github.com/servo/servo/blob/3f529b0025bd01c4fa82defbd147dab8d9cb0a68/components/script/dom/htmlselectelement.rs#L72)):</sup>
Since we implement [`Deref` on `Root`](http://doc.servo.org/servo/script/dom/bindings/js/struct.Root.html), you [don't need to explicitly](https://doc.rust-lang.org/nightly/book/deref-coercions.html#deref-and-method-calls) call `.r()`. It's the same reason one can do `String::from("hello").len()` even though `len()` is a `&str` method (`String` `Deref`s to `&str`)
<!-- Reviewable comment -K0JrEWd6-U2Hfn1thlo:702906460 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 78 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JoQIP2vSwpPy3Fr_I)** ([raw file](https://github.com/servo/servo/blob/3f529b0025bd01c4fa82defbd147dab8d9cb0a68/components/script/dom/htmlselectelement.rs#L78)):</sup>
Actually, `get_enabled_state` might be something different 
<!-- Reviewable comment -K0NiRESFnKjpnBz0ag2:-1119134365 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

dagnir commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 72 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JrEWd6-U2Hfn1thln)** ([raw file](https://github.com/servo/servo/blob/3f529b0025bd01c4fa82defbd147dab8d9cb0a68/components/script/dom/htmlselectelement.rs#L72)):</sup>
Removed unneeded calls to `r()`.
<!-- Reviewable comment -K0OsJ52lwkp7Oix1McY:89733294 -->

---

<sup>**[components/script/dom/htmlselectelement.rs, line 78 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7963#-K0JoQIP2vSwpPy3Fr_I)** ([raw file](https://github.com/servo/servo/blob/3f529b0025bd01c4fa82defbd147dab8d9cb0a68/components/script/dom/htmlselectelement.rs#L78)):</sup>
Simplified check.
<!-- Reviewable comment -K0OsGyaVfPkfdeQ57dm:951651177 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
FYI, the latest commit doesn't pass the style checker, one of the lines is too long

https://travis-ci.org/servo/servo/jobs/84851573

dagnir commented:
Hi @frewsxcv, is there anything else needed for this PR?

frewsxcv commented:
Have you tried running the web-platform-tests (`test-wpt`) test suite to see if any tests unexpectedly pass or fail. If you need help doing that, check out [this link](https://github.com/servo/servo/tree/master/tests/wpt#running-the-tests)). If after running the tests, nothing changes, a regression test will need to be written to make sure algorithm this doesn't break in the future. If you need help with any of this or want clarification, please let me know, I'll be happy to help! I'm currently doing IRL non-Servo work at the moment, but when I get home I'll actually read through the spec and make sure your implementation looks correct.

dagnir commented:
I'm running `test-wpt` now.  If you would show me how to get started with writing the regression test for this, I'd appreciate it!

Thanks!

frewsxcv commented:
(to anyone who is reading this PR and wondering why we haven't responded, we just chatted 1 on 1 on IRC)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7984) made this pull request unmergeable. Please resolve the merge conflicts.

dagnir commented:
@frewsxcv, I've added sow tests, please take a look.  Only the third one passes at the moment because `ask_for_reset` doesn't get called on node insert and remove.

eefriedman commented:
The relevant hooks for node insert and remove are `bind_to_tree()` and `unbind_from_tree()`.

dagnir commented:
Thanks!

On Sat, Oct 17, 2015, 3:33 PM eefriedman <notifications@github.com> wrote:

> The relevant hooks for node insert and remove are bind_to_tree() and
> unbind_from_tree().
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/servo/servo/pull/7963#issuecomment-148965829>.
>

frewsxcv commented:
FYI, in case you missed it, there are merge conflicts that will eventually need to be addressed

dagnir commented:
Branch is now up-to-date.

dagnir commented:
Hmm it seems `unbind_from_tree` is called after the node is already removed, so I can't iterate over the ancestors to find the `select` (from `HTMLOptionElement`).  Is there a "will remove" or similar hook I can use?

Also, looks like `html/semantics/forms/the-select-element/select-remove.html` has multiple failures atm, so maybe there's more work involved for removal to work?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7935) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
I wonder if we should be including the previous parent node as an argument or something. That wouldn't work for notifying children of the root unbound node, though. Do you have thoughts on this, @Ms2ger?

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
The spec doesn't say anything about 'display size' == 0, right?
<!-- Reviewable comment -K17OtqnqAvpbb-Z9FtD:1213041145 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 75 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17PPRaBIr01w5O-ABK)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L75)):</sup>
Instead of keeping track of the number of elements that are 'selected' (which is a number we don't really care about and might as well be a boolean), why not just check here if `last_selected.is_none()`
<!-- Reviewable comment -K17PPRaBIr01w5O-ABL:-191284952 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

eefriedman commented:
> The spec doesn't say anything about 'display size' == 0, right?

The display size of a select is always greater than zero.  That's the only reasonable way to interpret https://html.spec.whatwg.org/multipage/forms.html#concept-select-size.

dagnir commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
Yeah I changed it to include 0 because `Size()` defaults 0: https://github.com/servo/servo/blob/941c06eb99533c6c1d55b4ad3b51804badf48a23/components/script/dom/htmlselectelement.rs#L28.  I checked and Gecko also defaults to 0, even if one option is being displayed.  Don't know if this is a bug or not 
<!-- Reviewable comment -K184sB0jXFXIZM86UBs:18375091 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

dagnir commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
Actually I think this is fine if `Size() == 0` means it's absent...spec says

>  if the size attribute is absent, then the element's display size is 4 if the element's multiple content attribute is present, and 1 otherwise.
<!-- Reviewable comment -K18M3gRBX4dk0W5c_3f:594995010 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
So there are two possibilities here I see:

* `size` attribute is explicitly specified
  * From the spec:  

> "The size attribute, if specified, must have a value that is a valid non-negative integer **greater than zero**."

* `size` attribute is not specified
  * From the spec

> "If applying those rules to the attribute's value is not successful, or if the size attribute is absent, then the element's display size is 4 if the element's multiple content attribute is present, and **1 otherwise**"

I don't see a scenario where it's possible for it to be zero *from the spec* (ignore our codebase and Gecko codebase)
<!-- Reviewable comment -K19M13AgSgQikNL1Ffl:1912391072 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
Also, something to keep in mind that I forgot to clarify in my previous comment here. `size` and `display size` are two separate concepts. It appears to be possible for `size` to be `0`, but not `display size`. Considering it's a separate concept, I think it's appropriate to have a separate method that returns the `display size` (with a link to the `display size` portion of the spec adjacent to it). That method just returns the `size` attribute if it's not `0`, those default values otherwise
<!-- Reviewable comment -K19NgQn8Fu0gcU3tYyG:1803087544 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8115) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
> "The size attribute, if specified, must have a value that is a valid non-negative integer greater than zero."

I don’t have any context, but this sounds like an author requirements. Sooner or later, a some website will violate that requirement. However wrong this is, a browser has to handle it somehow. Another part of the spec should define how, through implementation requirements.

frewsxcv commented:
@dagnir Let me know if you need any assistance with this or if you need any clarification on the next steps

dagnir commented:
Thanks @frewsxcv.  I haven't had a change to work on this the past couple days, but I should have an updated patch up tonight or tomorrow.  I'll ping you on IRC if I have questions

dagnir committed:
Implement pick_option.

dagnir committed:
Update and correct tests.

dagnir committed:
ask for reset and pick on option insert.

dagnir commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 76 \[r6\]](https://reviewable.io:443/reviews/servo/servo/7963#-K17OtqnqAvpbb-Z9FtC)** ([raw file](https://github.com/servo/servo/blob/fb5f6c3169de9ace787b511889cf28f73cb0f2c0/components/script/dom/htmlselectelement.rs#L76)):</sup>
Done.
<!-- Reviewable comment -K1M0AN8z4DSQNam8bI3:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

dagnir committed:
Move cast into if block.

dagnir committed:
Update test.

Fix comments and test when size > 1.

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 108 \[r11\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1MK3rwehboMZm1Kq_a)** ([raw file](https://github.com/servo/servo/blob/fcfc3918a4e53931ef8cd61a1f6a51d9bfc2e20f/components/script/dom/htmlselectelement.rs#L108)):</sup>
The way you wrote this is fine, but you can also do:

```
match self.Size() {
  0 if self.Multiple() => 4,
  0 => 1,
  s => s,
}
```
<!-- Reviewable comment -K1MK3rwehboMZm1Kq_b:-230509190 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 82 \[r11\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1MJVz-efchWQhrLrCG)** ([raw file](https://github.com/servo/servo/blob/fcfc3918a4e53931ef8cd61a1f6a51d9bfc2e20f/components/script/dom/htmlselectelement.rs#L82)):</sup>
Using `unwrap` is generally a bad practice. A couple easy ways to avoid this:

```
if Some(last_slected) = last_selected {
  ...
} else {
  ...
}
```

or even better:

```
match last_selected {
  Some(last_selected) => ...,
  None => ...,
}
```
<!-- Reviewable comment -K1MJVz-efchWQhrLrCH:-999374581 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-ask-for-reset.html, line 39 \[r9\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1MmoOcVqJp6lSMMHn_)** ([raw file](https://github.com/servo/servo/blob/663801ed79eefc6d20511cade9efb6cdf6e905dc/tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-ask-for-reset.html#L39)):</sup>
Nit: extra indentation here
<!-- Reviewable comment -K1MmoOdsBFb6jb_bMAZ:-1338541199 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
Looks good to me after my latest comments! Thanks for all your hard work :)

dagnir committed:
Replace if-else with match.

dagnir committed:
Remove extra indent.

dagnir committed:
Mark failure for remove test as expected.

Currently unable to get former parent after unbind_from_tree, so it's
not possible to call ask_for_reset upon removal.

dagnir commented:
Thanks for all the help!.  Also FYI, I marked the first test as `expected: FAIL` since as discussed earlier in this thread, I can't get to the `select` in `unbind_from_tree` to call `ask_for_reset`.
<!-- Reviewable comment -K1SUGqqGg9Ds5U9zvVq:371781425 -->

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 82 \[r11\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1MJVz-efchWQhrLrCG)** ([raw file](https://github.com/servo/servo/blob/fcfc3918a4e53931ef8cd61a1f6a51d9bfc2e20f/components/script/dom/htmlselectelement.rs#L82)):</sup>
Done.
<!-- Reviewable comment -K1SUDeCjfyRI1yIJRKb:-499146009 -->

---

<sup>**[tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-ask-for-reset.html, line 39 \[r9\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1MmoOcVqJp6lSMMHn_)** ([raw file](https://github.com/servo/servo/blob/663801ed79eefc6d20511cade9efb6cdf6e905dc/tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-ask-for-reset.html#L39)):</sup>
Done.
<!-- Reviewable comment -K1SUEvpNc0YmDKu6sCE:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented on tests/wpt/metadata/html/semantics/forms/the-select-element/select-ask-for-reset.html.ini:
```diff
@@ -0,0 +1,4 @@
+[select-ask-for-reset.html]
```
Is there any reason why this test is marked as failing?

jdm replied:
See first paragraph of previous comment.

frewsxcv commented:
<sup>**[components/script/dom/htmloptionelement.rs, line 145 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1U7ZHePn1tEG8neqE4)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmloptionelement.rs#L145)):</sup>
Where in the spec does it talk about picking options upon setting `selected`?
<!-- Reviewable comment -K1U7ZHePn1tEG8neqE5:-844050363 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 85 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1U96ZKuqAf82o1BLWn)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L85)):</sup>
Does this method correspond to a particular section in the spec?
<!-- Reviewable comment -K1U96ZKuqAf82o1BLWo:-31220850 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

dagnir commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 85 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1U96ZKuqAf82o1BLWn)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L85)):</sup>
Yes, it's here https://html.spec.whatwg.org/multipage/forms.html#concept-select-pick (and the paragraph after it).  I'm not sure now whether I should have left it to be fixed in another issue, but the second test doesn't pass without it.
<!-- Reviewable comment -K1VSF7fOIAxTT6iNyXr:-755362311 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

jdm commented:
@dagnir With regards to the unbind_from_tree issue, you might be interested in the changes in https://github.com/servo/servo/compare/nox:finish-ranges .

dagnir commented:
@jdm awesome, I'll keep an eye out to see when it lands.

eefriedman commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmloptionelement.rs, line 147 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f6MLR0aExVK2S5Crt)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmloptionelement.rs#L147)):</sup>
Weird indentation.
<!-- Reviewable comment -K1f6MLT6AKpp4BeCVtA:-2021788707 -->

---

<sup>**[components/script/dom/htmloptionelement.rs, line 208 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f62_g3NGUR0srr6TT)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmloptionelement.rs#L208)):</sup>
This is missing a call to the ask-for-reset algorithm.  (Sort of an edge case... but consider what happens if you insert an option which isn't selected into an empty select.)
<!-- Reviewable comment -K1f62_g3NGUR0srr6TU:-171812195 -->

---

<sup>**[components/script/dom/htmlselectelement.rs, line 67 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f33GEuYH5nQAt2vVK)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L67)):</sup>
Whitespace.
<!-- Reviewable comment -K1f33GEuYH5nQAt2vVL:471607052 -->

---

<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
`if let Some(last_selected) = `...
<!-- Reviewable comment -K1f3QWPdmQYoDX01bxD:-1623805723 -->

---

<sup>**[components/script/dom/htmlselectelement.rs, line 85 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1U96ZKuqAf82o1BLWn)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L85)):</sup>
It would be nice to include that URL as a comment before the definition of this function.
<!-- Reviewable comment -K1f0r3MfPIb_JJuP-hJ:-843473536 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

dagnir committed:
Add fixes based on review.

- Whitespace and indentation issues

- call as_for_reset on option insert

- add link to 'pick' in standard

dagnir commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmloptionelement.rs, line 147 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f6MLR0aExVK2S5Crt)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmloptionelement.rs#L147)):</sup>
Done.
<!-- Reviewable comment -K1fE76Icbc5WEyE_KH-:-499146009 -->

---

<sup>**[components/script/dom/htmloptionelement.rs, line 208 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f62_g3NGUR0srr6TT)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmloptionelement.rs#L208)):</sup>
Done.
<!-- Reviewable comment -K1fEi2fU0FR9KADww5B:-499146009 -->

---

<sup>**[components/script/dom/htmlselectelement.rs, line 67 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f33GEuYH5nQAt2vVK)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L67)):</sup>
Done.
<!-- Reviewable comment -K1fD2_bHTDDyxkB7uVt:-499146009 -->

---

<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
From what I understood from @frewsxcv's comment here: https://reviewable.io/reviews/servo/servo/7963#-K1MJVz-efchWQhrLrCG, this was preferable, unless I misunderstood 
<!-- Reviewable comment -K1fCPr_IZAhcYzpS9Se:199720349 -->

---

<sup>**[components/script/dom/htmlselectelement.rs, line 85 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1U96ZKuqAf82o1BLWn)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L85)):</sup>
Done.
<!-- Reviewable comment -K1fEd5AuZYqgnrgkso3:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

eefriedman commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
`if let` is generally preferred over `is_none()`+`unwrap()`; I don't think that review comment was specifically suggesting the opposite.
<!-- Reviewable comment -K1fJUuYcnjRF29TdutW:-1250571689 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

dagnir commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
Hmm is it possible you saw an outdated revision?  This is what it currently looks like https://github.com/dagnir/servo/commit/b5e991c89e11be2467b30d1212052c57e21c81d6
<!-- Reviewable comment -K1fL52oVHK2qfVrDrCD:454031287 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

eefriedman commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
Oh, right, sorry, I somehow got flipped to the wrong revision just now.  Still, `if let` is generally preferred over match.
<!-- Reviewable comment -K1fM5QD3EoyN9eFgPob:1461695455 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
<sup>**[components/script/dom/htmlselectelement.rs, line 73 \[r14\]](https://reviewable.io:443/reviews/servo/servo/7963#-K1f3QWOC3L1VnCL-ip3)** ([raw file](https://github.com/servo/servo/blob/15a8b6b62c6129711d37f02c4577f7448f679998/components/script/dom/htmlselectelement.rs#L73)):</sup>
"Oh, right, sorry, I somehow got flipped to the wrong revision just now. Still, if let is generally preferred over match."

I disagree, I think they're both perfectly valid. `match` gives you the benefit of erroring when all cases are not covered.
<!-- Reviewable comment -K1fNHLRSBaHsGiRMyvB:-1377702370 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

frewsxcv commented:
Thanks for looking at this @eefriedman, I've been busy these past few days and haven't had time to review this.

eefriedman commented:
> I disagree, I think they're both perfectly valid. match gives you the benefit of erroring when all cases are not covered.

You're right.  I'll narrow it a bit; "if let" is preferred over "match" for `Option<T>`.

dagnir commented:
@frewsxcv @eefriedman okay to leave, or should I change it?

eefriedman commented:
Please change it.  Also, please refactor the identical code in `SetSelected` and `ask_for_reset`.

dagnir commented:
@eefriedman I'm not sure what identical code you're referring to; unless you mean the option element traversal in `ask_for_reset` and `pick_option`?

eefriedman commented:
SetSelected:

```
+        if let Some(select) = self.upcast::<Node>().ancestors()
+                .filter_map(Root::downcast::<HTMLSelectElement>).next() {
+            if selected {
+                select.pick_option(self);
+            }
+            select.ask_for_reset();
+        }
```

ask_for_reset:

```
+        if let Some(select) = self.upcast::<Node>().ancestors()
+                .filter_map(Root::downcast::<HTMLSelectElement>)
+                .next() {
+            if self.Selected() {
+                select.pick_option(self);
+            }
+            select.ask_for_reset();
+        }
```

eefriedman commented:
Err, oops, I meant SetSelected and bind_to_tree.

dagnir committed:
Add fixes based on review.

- Use if let instead of match for Option

- Refactor common code into pick_if_selected_and_reset

dagnir commented:
Thanks, done.
<!-- Reviewable comment -K1g1yTENftceCSU_qpN:1457700874 -->

---

Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7963)
<!-- Sent from Reviewable.io -->

eefriedman commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 4849033 has been approved by `eefriedman`

<!-- @bors-servo r=eefriedman 4849033297815625939d2f7d8f68a14adfaa5477 -->

bors-servo commented:
:hourglass: Testing commit 4849033 with merge 3951c57...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/73)

frewsxcv commented:
@bors-servo retry

* https://github.com/servo/servo/issues/8236

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/71), [gonk](http://build.servo.org/builders/gonk/builds/72), [linux-dev](http://build.servo.org/builders/linux-dev/builds/65), [linux-rel](http://build.servo.org/builders/linux-rel/builds/65), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/72) are reusable. Rebuilding only [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/73), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/78)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/71), [gonk](http://build.servo.org/builders/gonk/builds/72), [linux-dev](http://build.servo.org/builders/linux-dev/builds/65), [linux-rel](http://build.servo.org/builders/linux-rel/builds/65), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/72), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/74), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/78)

Ms2ger commented:
For future reference:

> "if let" is preferred over "match" for `Option<T>`

is not true. It is preferred only for (some) cases where only a single variant has code associated with it.
